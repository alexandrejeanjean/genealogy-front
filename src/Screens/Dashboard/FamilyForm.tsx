import React, { PureComponent } from 'react'
import { Form, Button, Modal } from 'react-bootstrap'

type Props = {
  handleSubmit: Function
  errorMsg: string
  show: boolean
  onHide: Function
}

interface State {
  name: string
}

class FamilyForm extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      name: '',
    }
  }

  handleChange = (event: React.FormEvent<{ name: string; value: string }>) => {
    const name = event.currentTarget.name
    const value = event.currentTarget.value
    this.setState(({ [name]: value } as unknown) as Pick<State, keyof State>)
  }

  _handleSubmit = (event: React.FormEvent<HTMLInputElement>) => {
    const { onHide } = this.props
    event.preventDefault()
    const { handleSubmit } = this.props
    handleSubmit(this.state)
    onHide()
  }

  render() {
    const { errorMsg, ...props } = this.props
    return (
      <Modal
        size='lg'
        aria-labelledby='contained-modal-title-vcenter'
        centered
        {...props}
      >
        <Modal.Header closeButton>
          <Modal.Title id='contained-modal-title-vcenter'>
            New family
          </Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form onSubmit={this._handleSubmit}>
            <Form.Group controlId='familyName'>
              <Form.Label>Name</Form.Label>
              <Form.Control
                type='text'
                name='name'
                placeholder='Enter family name'
                onChange={this.handleChange}
              />
            </Form.Group>
            <Button type='submit' className='btn-primary mt-3'>
              Create family
            </Button>
          </Form>
          {errorMsg && <p className='error-text'>{errorMsg}</p>}
        </Modal.Body>
      </Modal>
    )
  }
}

export default FamilyForm
