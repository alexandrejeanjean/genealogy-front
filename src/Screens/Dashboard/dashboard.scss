.new-item-btn {
  width: 200px;
  margin-left: 2.5%;
}
