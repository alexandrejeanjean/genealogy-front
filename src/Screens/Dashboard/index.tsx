import React, { useEffect, useState } from "react";
import { useHistory } from "react-router-dom";
import Main from "../Theme/index";
import apiClient from "../../api";
import ListWrapper from "../../SharedComponents/ListWrapper/ListWrapper";
import CreateItemBtn from "../../SharedComponents/CreateItemBtn/CreateItemBtn";
import ModalForm from "../../SharedComponents/ModalForm/ModalForm";
import family from "../../assets/family.svg";
import "./dashboard.scss";

type TFamily = {
  id: number;
  name: string;
};

type TGetFamilies = TFamily[];

const Dashboard = () => {
  const [list, setList] = useState<TGetFamilies>([]);
  const [modalShow, setModalShow] = useState<boolean>(false);
  const [errorMsg, setErrorMsg] = useState("");
  const history = useHistory();

  // Get family datas ---------------------------------------
  const getFamilies = async () => {
    try {
      const response = await apiClient.get("/families");
      const families = response.data;
      if (families) {
        setList(families);
      }
    } catch (err) {
      if (err && err.response) {
        const axiosError = err;
        console.log("Error :: get families", axiosError.response);
        return axiosError.response.data;
      }

      throw err;
    }
  };

  // Get datas on component did mount --------------------------
  useEffect(() => {
    getFamilies();
  }, []);

  // Add family  ---------------------------------------
  const createFamily = async (family: { name: string }) => {
    try {
      const response = await apiClient.post(`/families`, { name: family.name });
      if (response) getFamilies();
    } catch (err) {
      if (err && err.response) {
        const axiosError = err;
        console.log("Error :: create family", axiosError.response);
        setErrorMsg(
          "Sorry, an error occured. Please try again or contact us if problem persists."
        );
        return axiosError.response.data;
      }
      throw err;
    }
  };

  // Delete family  ---------------------------------------
  const deleteFamily = async (family: TFamily) => {
    const familyId: number = family.id;
    try {
      const response = await apiClient.delete(`/families/${familyId}`);
      if (response) getFamilies();
    } catch (err) {
      if (err && err.response) {
        const axiosError = err;
        console.log("Error :: delete family", axiosError.response);
        return axiosError.response.data;
      }
      throw err;
    }
  };

  const goTo = (familyId: number) => {
    return history.push({
      pathname: "/family",
      state: { datas: familyId },
    });
  };

  return (
    <Main pageTitle="Dashboard">
      <section>
        <ListWrapper
          datas={list}
          getAvatar={() => family}
          handleClick={(familyId: number) => goTo(familyId)}
          handleDelete={(family: TFamily) => deleteFamily(family)}
        >
          <CreateItemBtn handleClick={() => setModalShow(true)} />
        </ListWrapper>

        <ModalForm
          title="Family"
          show={modalShow}
          inputs={[
            { name: "name", placeholder: "Ex: Armstrong", inputType: "text" },
          ]}
          onHide={() => setModalShow(false)}
          handleSubmit={createFamily}
          errorMsg={errorMsg}
        />
        {errorMsg && <p className="error-text ml-5">{errorMsg}</p>}
      </section>
    </Main>
  );
};

export default Dashboard;
