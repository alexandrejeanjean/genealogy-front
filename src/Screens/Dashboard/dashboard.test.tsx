import React from "react";
import { shallow } from "enzyme";
import Dashboard from "./index";

const DashboardProps = {
  setToastVisible: jest.fn(),
};

describe("Dashboard", () => {
  it("should render correctly", () => {
    const Component = shallow(<Dashboard {...DashboardProps} />);
    expect(Component).toMatchSnapshot();
  });
});
