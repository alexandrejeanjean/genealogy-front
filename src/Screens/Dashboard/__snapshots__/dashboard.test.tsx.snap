// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
